src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.Entry.Roundtrip.Tests.cs

```diff
@@ -0,0 +1,137 @@
+                    yield return new object[] { TarEntryFormat.V7, v7EntryType, name };
+                }
+
+                foreach (string name in GetNamesNonAsciiTestData(NameCapabilities.NameAndPrefix).Concat(GetNamesPrefixedTestData(NameCapabilities.NameAndPrefix)))
+                {
+                    yield return new object[] { TarEntryFormat.Ustar, entryType, name };
+                }
+
+                foreach (string name in GetNamesNonAsciiTestData(NameCapabilities.Unlimited).Concat(GetNamesPrefixedTestData(NameCapabilities.Unlimited)))
+                {
+                    yield return new object[] { TarEntryFormat.Pax, entryType, name };
+                    yield return new object[] { TarEntryFormat.Gnu, entryType, name };
+                }
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(NameRoundtripsTheoryData))]
+        public void NameRoundtrips(TarEntryFormat entryFormat, TarEntryType entryType, string name)
+        {
+            TarEntry entry = InvokeTarEntryCreationConstructor(entryFormat, entryType, name);
+            entry.Name = name;
+
+            MemoryStream ms = new();
+            using (TarWriter writer = new(ms, leaveOpen: true))
+            {
+                writer.WriteEntry(entry);
+            }
+
+            ms.Position = 0;
+            using TarReader reader = new(ms);
+
+            entry = reader.GetNextEntry();
+            Assert.Null(reader.GetNextEntry());
+            Assert.Equal(name, entry.Name);
+        }
+
+        public static IEnumerable<object[]> LinkNameRoundtripsTheoryData()
+        {
+            foreach (TarEntryType entryType in new[] { TarEntryType.SymbolicLink, TarEntryType.HardLink })
+            {
+                foreach (string name in GetNamesNonAsciiTestData(NameCapabilities.Name).Concat(GetNamesPrefixedTestData(NameCapabilities.Name)))
+                {
+                    yield return new object[] { TarEntryFormat.V7, entryType, name };
+                    yield return new object[] { TarEntryFormat.Ustar, entryType, name };
+                }
+
+                foreach (string name in GetNamesNonAsciiTestData(NameCapabilities.Unlimited).Concat(GetNamesPrefixedTestData(NameCapabilities.Unlimited)))
+                {
+                    yield return new object[] { TarEntryFormat.Pax, entryType, name };
+                    yield return new object[] { TarEntryFormat.Gnu, entryType, name };
+                }
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(LinkNameRoundtripsTheoryData))]
+        public void LinkNameRoundtrips(TarEntryFormat entryFormat, TarEntryType entryType, string linkName)
+        {
+            string name = "foo";
+            TarEntry entry = InvokeTarEntryCreationConstructor(entryFormat, entryType, name);
+            entry.LinkName = linkName;
+
+            MemoryStream ms = new();
```

Comment:
This test also needs to be extended to unseekable archives. might have to do it in [my other PR for extended attributes](https://github.com/dotnet/runtime/pull/75237), because I suspect this test would fail in this PR.

The GNU test is the one that would not fail if you extend it in this PR. That format saves the long link in the data section of another entry.

Reply:
I just added tests for unseekable streams as well.

Reply:
Thanks. Surprised that there are no exceptions being thrown.